aws/resource_aws_kms_external_key.go

```diff
@@ -0,0 +1,481 @@
+	})
+
+	if isResourceTimeoutError(err) {
+		getParametersForImportOutput, err = conn.GetParametersForImport(getParametersForImportInput)
+	}
+
+	if err != nil {
+		return fmt.Errorf("error getting parameters for import: %s", err)
+	}
+
+	if getParametersForImportOutput == nil {
+		return fmt.Errorf("error getting parameters for import: empty response")
+	}
+
+	keyMaterial, err := base64.StdEncoding.DecodeString(keyMaterialBase64)
+
+	if err != nil {
+		return fmt.Errorf("error Base64 decoding key material: %s", err)
+	}
+
+	publicKey, err := x509.ParsePKIXPublicKey(getParametersForImportOutput.PublicKey)
+
+	if err != nil {
+		return fmt.Errorf("error parsing public key: %s", err)
+	}
+
+	encryptedKeyMaterial, err := rsa.EncryptOAEP(sha256.New(), rand.Reader, publicKey.(*rsa.PublicKey), keyMaterial, []byte{})
+
+	if err != nil {
+		return fmt.Errorf("error encrypting key material: %s", err)
+	}
+
+	importKeyMaterialInput := &kms.ImportKeyMaterialInput{
+		EncryptedKeyMaterial: encryptedKeyMaterial,
+		ExpirationModel:      aws.String(kms.ExpirationModelTypeKeyMaterialDoesNotExpire),
+		ImportToken:          getParametersForImportOutput.ImportToken,
+		KeyId:                aws.String(keyID),
+	}
+
+	if validTo != "" {
+		t, err := time.Parse(time.RFC3339, validTo)
+
+		if err != nil {
+			return fmt.Errorf("error parsing valid to timestamp: %s", err)
+		}
+
+		importKeyMaterialInput.ExpirationModel = aws.String(kms.ExpirationModelTypeKeyMaterialExpires)
+		importKeyMaterialInput.ValidTo = aws.Time(t)
+	}
+
+	// Handle KMS eventual consistency
+	err = resource.Retry(1*time.Minute, func() *resource.RetryError {
+		_, err := conn.ImportKeyMaterial(importKeyMaterialInput)
+
+		if isAWSErr(err, kms.ErrCodeNotFoundException, "") {
+			return resource.RetryableError(err)
+		}
+
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+
+		return nil
+	})
+
+	if isResourceTimeoutError(err) {
+		_, err = conn.ImportKeyMaterial(importKeyMaterialInput)
+	}
+
+	if err != nil {
+		return fmt.Errorf("error importing key material: %s", err)
+	}
+
+	return nil
+}
+
+func waitForKmsKeyScheduleDeletion(conn *kms.KMS, keyID string) error {
+	// Wait for propagation since KMS is eventually consistent
+	input := &kms.DescribeKeyInput{
+		KeyId: aws.String(keyID),
+	}
+
+	wait := resource.StateChangeConf{
+		Pending:                   []string{kms.KeyStateDisabled, kms.KeyStateEnabled},
+		Target:                    []string{kms.KeyStatePendingDeletion},
+		Timeout:                   20 * time.Minute,
+		MinTimeout:                2 * time.Second,
+		ContinuousTargetOccurence: 10,
+		Refresh: func() (interface{}, string, error) {
+			output, err := conn.DescribeKey(input)
+
+			if isAWSErr(err, kms.ErrCodeNotFoundException, "") {
+				return 42, kms.KeyStatePendingDeletion, nil
+			}
+
+			if err != nil {
+				return nil, kms.KeyStateUnavailable, err
+			}
+
+			if output == nil || output.KeyMetadata == nil {
+				return 42, kms.KeyStatePendingDeletion, nil
```

Comment:
We discussed this one out of band as it had confused me when I was first working with the codebase, but it is a fairly common pattern and does indeed deserve its own constant. If error is `nil`, the return value for the API object is to prevent `StateChangeConf` from repeatedly retrying to "find" the API object. The choice of `42` is a Hitchhiker's Guide reference.

Reply:
Since speaking about this I've come across a few other occurrences :rofl: . Thanks for the background on the value and its usage. 

Not for this PR but we should make it a constant for sure. In keeping with the theme here's a of couple of suggestions :stuck_out_tongue_winking_eye: 

```go
const AnswerToLife = 42
const HGConstant = 42
```